tools/topology/topology2/include/components/gain.conf

```diff
@@ -152,6 +152,41 @@ Class.Widget."gain" {
 				}
 			}
 		}
+
+		# mute switch control
+		mixer."2" {
```

Comment:
Yes, the switch is 4ch while volume is 2ch. It looks weird. In the other hand, the dmic gain widget is 4ch while volume is still 2ch. I don't know if there is a elegant solution for it so I just compare the channel count of gain widget and the switch from host side to determine how many channels to update.

Reply:
Hmm, I'm just concerned this is defined in the generic component definition. In tplg1, we different pipeline templates for pipes with volume and volume-switch topologies. Additionally, e.g. the channel map (and count) was defined in the platform definition, like topology1/platform/intel/intel-generic-dmic.m4.

If we define a 4ch switch here, it will be in all PGA/gain, in all pipelines for all targets. 

@ranj063 @jsarha @singalsu any examples how to do the above with tplg2?

Reply:
Its fine defining the 4 channels here and ALSA will still display them. However, unused channels will be immutable and will show as unmuted but cant be changed. i.e. ```snd_ctl_elem_write()``` will return 0 (success but value not changed). This means the default channels mute state would need to be muted for unused channels.

Reply:
I think similar update should be done for above mixer."1" widget. Currently in e.g. sof-hda-generic-4h.tplg it's not possible to control channels 3-4 volume.

Reply:
I just tried this with sof-hda-generic-2ch.tplg. There is 4ch switch control but it works logically:

```
singalsu@ronsu:~$ amixer -c1 cset name='Dmic0 Capture Switch' off
numid=44,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=4
  : values=off,off,off,off
singalsu@ronsu:~$ amixer -c1 cset name='Dmic0 Capture Switch' on
numid=44,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=4
  : values=on,on,on,on
singalsu@ronsu:~$ amixer -c1 cset name='Dmic0 Capture Switch' on,off
numid=44,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=4
  : values=on,off,on,on
singalsu@ronsu:~$ amixer -c1 cset name='Dmic0 Capture Switch' off,on
numid=44,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=4
  : values=off,on,on,on

```

Reply:
Ack, this is fine. If the mixer object is only added to tplgs which set a non-NULL name, so the switch is only added to dmic pipeline (and not to all pipelines that use the Gain widget class). Not very intuitive, but this is documented in mixer.conf.